### PR TITLE
Request TLS certificate for tailnet domain alias.

### PR DESCRIPTION
**Pull request by ananthb**
# Proposed Changes

Request TLS certificate from tailscale. HTTPS must be enabled for the tailnet.
https://tailscale.com/kb/1153/enabling-https/

## Related Issues

https://github.com/hassio-addons/addon-tailscale/issues/62

@lmagyar has an even easier solution in #137.

### CODE REVIEW

**Comment by ananthb** [comment]
`tailscale cert` has to be run periodically to keep certs fresh. I can either run a background script from the post script, or start a cron daemon and use that. Any suggestions here?

**Review by franzbischoff — COMMENTED** [review]
Right!
1. add ssl folder to the addon
2. add options
3. fetch new cert every day

👍

**Comment by ananthb** [comment]
Yep that's what it does. The cert process runs once every 24 hours to fetch a new cert. The SSL folder is mapped rw. There is an option to enable the addon and a mandatory one for the tailnet domain name.

**Comment by franzbischoff** [comment]
The PR is failing because of the target name :-/

**Comment by frenck** [comment]
> The PR is failing because of the target name :-/

Its not.

**Comment by ananthb** [comment]
Yes automation seems to expect a label but the PR is currently unlabeled. I don't have permission to label it either.

**Comment by frenck** [comment]
Yup, that is not a issue

**Comment by ananthb** [comment]
@frenck so how do we proceed?

**Comment by frenck** [comment]
By having patience

**Comment by ananthb** [comment]
@frenck is this failure unexpected? I don't understand the error message.

**Comment by frenck** [comment]
>  is this failure unexpected?

Yes

> I don't understand the error message.

OK... not sure what to do about that 🤷 It complains a label is missing, so that needs to be added by someone with permissions to do that.

**Comment by ananthb** [comment]
I meant after you added the label, the amd64 build failed at a docker buildx step. I was wondering if i had to fix something but that doesn't appear to be the case.

File: tailscale/config.yaml
Comment (frenck) [inline tailscale/config.yaml:25]: Maybe we can just default these and remove the config?

For example by storing them in `/ssl/tailscale/[fullchain|privkey].pem`?

File: tailscale/config.yaml
Comment (frenck) [inline tailscale/config.yaml:35]: Maybe we don't need `request_cert` at all? For example, if one provides the domain alias, that should already be enough?

File: tailscale/rootfs/etc/services.d/cert/run
Comment (frenck) [inline tailscale/rootfs/etc/services.d/cert/run:10]: we don't have to check this constantly (every 24 hours). If this is kept, we can check it once (in a `cont-init` script).

**Review by frenck — CHANGES_REQUESTED** [review]

File: tailscale/config.yaml
Comment (franzbischoff) [reply tailscale/config.yaml:25]: I'm using as @frenck says. IMHO it is safer to have a different folder for tailscale as you may still need the non-tailscale certificates for something else. You can define which TSL the services will use, by-service, instead of forcing all [working] services to use tailscale's and for some reason all breaks.

An use case I can provide is the inability to login to HA itself just because I was relying on tailscale, and the tailscale addon went down.

my 2 cents

File: tailscale/config.yaml
Comment (franzbischoff) [reply tailscale/config.yaml:35]: And if we want to temporarily stop renewal?

File: tailscale/rootfs/etc/services.d/cert/run
Comment (franzbischoff) [reply tailscale/rootfs/etc/services.d/cert/run:10]: Can we at least auto-check every week? Or, grab the expire date and schedule it in someway. I know we frequently have updates and restart all services (and TS would try to renew the certificates), but some users or at some period we could have it untouched for a long time.

Anyway, if the certificate expires the option for the user is to restart the container, right?

File: tailscale/config.yaml
Comment (frenck) [reply tailscale/config.yaml:35]: That is a shallow use case to justify it imho. If wanted, one could temporary remove the alias as well 🤷

File: tailscale/config.yaml
Comment (ananthb) [reply tailscale/config.yaml:25]: I'd like the option to choose, because I'm going to run HTTPS only on the tailscale address and firewall everything else. We could change the default to something else but i went with what other certificate add-ons are doing.

File: tailscale/config.yaml
Comment (ananthb) [reply tailscale/config.yaml:35]: Yeah makes sense. We can do without request_cert.

File: tailscale/rootfs/etc/services.d/cert/run
Comment (ananthb) [reply tailscale/rootfs/etc/services.d/cert/run:10]: Hmm. The issue is renewal. If i have a long uptime without updates or restarts i still want new certs to be fetched automatically. The bash sleep is super light weight so i went with it.

File: tailscale/config.yaml
Comment (franzbischoff) [reply tailscale/config.yaml:35]: @frenck yes, it is a shallow use case, I'm fine with removing the `request_cert`. My "concern" was that if someone uses a complex regex, he would have to save it somewhere. But anyway, still shallow :-)

Comment (ananthb) [reply tailscale/config.yaml:35]: Users don't input regex though. The regex validates the config option which expects a domain name in the form xyz.ts.net where xyz is your tailnet domain alias from the tailscale admin.

File: tailscale/config.yaml
Comment (frenck) [reply tailscale/config.yaml:24]: > I'd like the option to choose

I'd like to reduce complexity and remove it.

Comment (ananthb) [reply tailscale/config.yaml:24]: Fair enough. I'll remove it.

**Comment by franzbischoff** [comment]
Not staled!

**Comment by martusi61** [comment]
I'm waiting too!

**Comment by nsiicm0** [comment]
@frenck I know it is not on us to accelerate the process, but could you have a look again at this PR? AFAIK, it has been sitting now for 3 months and all the changes you requested have been implemented. This feature would be a real game changer for the addon.

Thank you so much!

**Comment by frenck** [comment]
I am abroad for a couple of weeks, so can't do anything now. Marked it to look at when I get back.

**Comment by ShadowJonathan** [comment]
(Not stale)

**Comment by ShadowJonathan** [comment]
(Still not stale)

**Comment by ananthb** [comment]
@frenck did you have any time to look at this? I'm starting to feel that it's taking so long because it directly competes with hassio's paid offering.

If you're not interested in this change, I'm more than happy to maintain a more up to date fork.

**Comment by ShadowJonathan** [comment]
(Not stale)

**Comment by ananthb** [comment]
@frenck this change is eagerly awaited by me and quite a few others it would seem. Any chance it gets merged soon?

If you don't want this in change please let us know. It's pretty inconsiderate to just leave us hanging.

**Comment by ShadowJonathan** [comment]
Anecdotally, I'm also awaiting this change as I've moved over every homelab-hosted service to tailscale HTTPS a few months ago, save for Home Assistant, because of this PR's timeline, currently I'm still using cloudflare and a port-forward instead.

**Comment by ananthb** [comment]
For what it's worth, I've stopped using homeassistant entirely because of the petty handling of this change. I will keep this PR open until they decide to do something about it.

**Comment by lmagyar** [comment]
Until it is merged I forked the repo, updated it to Tailscale v1.34.2, and bumped the add-on base image also to v13.0.1.

And I merged this PR and made some modifications and bugfixes:
- the Configuration option is `cert_domain` (not `domain_alias`), it is much more logical, it is analog to the `tailscale cert <domain>` command
- it creates the `/ssl/tailscale` directory before fetching the cert
- uses `--statedir=/data` instead of `--state=/data/tailscaled.state`, see https://github.com/tailscale/tailscale-qpkg/pull/64
- fetch certificate on start and only within 1 week before expiration
- modified the daemon finish script to the usual script, I know new add-ons have better, it can be bumped later with other scripts

See: https://github.com/lmagyar/homeassistant-addon-tailscale

Disclaimer: I'm just a tinkerer, and "it works on my machine..." (tested on armv7)

I will create some new PR-s soon.

**Comment by franzbischoff** [comment]
Hi

My domain is from the beginning and doesnt comply with your regex filter...
I use only letters and dashes for ex, and the regex fails...
--
Francisco Bischoff


Laszlo Magyar ***@***.***> escreveu no dia sábado, 7/01/2023
à(s) 21:38:

**Comment by lmagyar** [comment]
> Hi My domain is from the beginning and doesnt comply with your regex filter...
> I use only letters and dashes for ex, and the regex fails...

Could you please send what is in your config? `cert_domain: ?????`

Maybe you entered the machine name also, like `cert_domain: abcd.tail1234.ts.net`, instead of only `cert_domain: tail1234.ts.net`

**Comment by franzbischoff** [comment]
I found that what you ask is the main domain: xyz.ts.net
I was trying the device subdomain: myha.xyz.ts.net

it works as it was :)
sorry

**Comment by sirreal** [comment]
I'm still hoping this progresses. Not stale.

**Comment by lmagyar** [comment]
I think #137 is a better solution, because that doesn't require any configuration and also provides a built-in proxy by tailscaled.

**Comment by ananthb** [comment]
@lmagyar that does look better!

I'm not holding my breath waiting for either one to get merged though. The response from Nabu Casa is pretty clear now.

> By having patience

@frenck this patient enough for you, or should we continue waiting forever?

Honestly this single issue has really turned me off of homeassistant. I'm a big believer in the spirit behind open source and Nabu Casa has shown absolutely none of that here. The community should really question their commitment to all of the projects they control.

**Pull request closed without merging**
